src/esperanza/finalizationstate.h

```diff
@@ -189,6 +189,9 @@ class FinalizationState : public FinalizationStateData {
   //! \brief Returns true if finalizer can vote in current dynasty
   bool IsFinalizerVoting(const Validator &finalizer) const;
 
+  //! \brief Calculate the epoch at which finalizer can start withdrawing
+  uint32_t CalculateWithdrawEpoch(const Validator &finalizer, Result *result_out) const;
```

Comment:
I have a feeling that a more consistent way would be:
`Result CalculateWithdrawEpoch(const Validator &finalizer, uint32_t* out_epoch) const;`
This is a more popular pattern where you return a value indicating success/failure and in case of success modify `out_*` parameter. With this you also don't need to bother with `std::numeric_limits<uint32_t>::max()`

Reply:
The reason it always returns epoch number and `Result` optionally is that in several places we don't care about the `Result`. 
The implementation of this function is to return the maximum `uint32_t` for `epoch` if it can't be computed. This allows writing the following code:
```c++
if (state.GetCurrentEpoch() < state.CalculateWithdrawEpoch(*validator,
                                                             /*result_out*/ nullptr)) {
    return ValidatorState::Phase::WAITING_FOR_WITHDRAW_DELAY;
  }
``` 
We use already this approch when we return `end_dynasty`, it's the max `uint32_t` when logout tx wasn't issued.

`Result` is used only in the validation, to be able to write better logs. In RPC/walletextention it is ignored.

WDYS?

Reply:
I've seen this usage. But I think it still confuses the reader about the function behaviour.
Function has two outcomes: 
1) It correctly computed withdraw epoch
2) Something went wrong
The fact that in 2 you will return maximum value is quite hidden thing. You need to read docs/source to figure it out. Plus it really hides the fact that something went wrong. 

And since you mentioned other functions - we have `CalculateWithdrawAmount` few lines above which has proposed signature.

Reply:
@AM5800 I understand your idea.
I want to point to one thing which wasn't clear in the previous comment before making changes.

This function *always* computes the correct withdraw epoch (there are no case when something went wrong). It will be either the epoch that will be reached soon or *infinity* (via max `uint32_t` that will take ~55 233 years to reach). And the user of this function should rely on that number (not on the `Result`)

Then there is the `Result` which can be used (if needed) to figure out are we trying to reach that *infinity* value or the one that was set when we did logout. You can also use a different approach to figure that out (if we logged out), check what is set in `end_dynasty`.

WDYS?

Reply:
I would maybe wait for more opinions

Reply:
@AM5800 I removed `result_out` parameter and unified the log message. We now use only returned `uint32_t` value.